sklearn/datasets/tests/test_svmlight_format.py

```diff
@@ -200,67 +201,90 @@ def test_invalid_filename():
-                                   zero_based=zero_based)
-                f.seek(0)
-
-                comment = f.readline()
-                try:
-                    comment = str(comment, "utf-8")
-                except TypeError:  # fails in Python 2.x
-                    pass
-
-                assert_in("scikit-learn %s" % sklearn.__version__, comment)
-
-                comment = f.readline()
-                try:
-                    comment = str(comment, "utf-8")
-                except TypeError:  # fails in Python 2.x
-                    pass
-
-                assert_in(["one", "zero"][zero_based] + "-based", comment)
-
-                X2, y2 = load_svmlight_file(f, dtype=dtype,
-                                            zero_based=zero_based)
-                assert_equal(X2.dtype, dtype)
-                assert_array_equal(X2.sorted_indices().indices, X2.indices)
-                if dtype == np.float32:
-                    assert_array_almost_equal(
+    X_sliced = X_sparse[np.arange(X_sparse.shape[0])]
+    y_sliced = y_sparse[np.arange(y_sparse.shape[0])]
+
+    for X in (X_sparse, X_dense, X_sliced):
+        for y in (y_sparse, y_dense, y_sliced):
+            for zero_based in (True, False):
+                for dtype in [np.float32, np.float64, np.int32]:
+                    f = BytesIO()
+                    # we need to pass a comment to get the version info in;
+                    # LibSVM doesn't grok comments so they're not put in by
+                    # default anymore.
+
+                    # make y conforms to shape: (n_samples, n_labels)
+                    if (sp.issparse(y) and y.shape[0] == 1):
+                        y = y.T
+
+                    dump_svmlight_file(X.astype(dtype), y, f, comment="test",
+                                    zero_based=zero_based)
+                    f.seek(0)
+
+                    comment = f.readline()
+                    try:
+                        comment = str(comment, "utf-8")
+                    except TypeError:  # fails in Python 2.x
+                        pass
+
+                    assert_in("scikit-learn %s" % sklearn.__version__, comment)
+
+                    comment = f.readline()
+                    try:
+                        comment = str(comment, "utf-8")
+                    except TypeError:  # fails in Python 2.x
+                        pass
+
+                    assert_in(["one", "zero"][zero_based] + "-based", comment)
+
+                    X2, y2 = load_svmlight_file(f, dtype=dtype,
+                                                zero_based=zero_based)
+                    assert_equal(X2.dtype, dtype)
+                    assert_array_equal(X2.sorted_indices().indices, X2.indices)
+
+                    X2_dense = X2.toarray()
+                    if y2.ndim == 1:
+                        y2_dense = sp.csr_matrix(y2).data
+                    else:
+                        y2_dense = sp.csr_matrix(y2).toarray()
```

Comment:
Oh since original code test X's rounding error when it is sparse and calls `toarray()`
```python
assert_array_almost_equal(
       X_dense.astype(dtype), X2.toarray(), 4)
```

I thought y must have the same test.

What I want to test is whether `y_sparse.toarray()` will almost equal to `y2`.
Maybe I should change `y2_dense` to `y_sparse.toarray()` to make it more clear?
Or you think these tests are unnecessary?